Add MatMulNBits Operator

georgen117 opened the pull request:
Description
This adds MatMulNBits to the ONNX standard.

Motivation and Context
Standardize Operators that are showing up in key LLM models.

[MatMulNBits](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.MatMulNBits) is already been defined in ONNXRuntime.

MatMulNBits node is already generated by "ONNX Runtime GenAI Model Builder" and "Intel Neural Compressor"

It can be found used in the int-4 quantized Phi-3 model hosted on [hugging face](https://huggingface.co/microsoft/Phi-3-mini-4k-instruct-onnx)

This specification differs from the com.microsoft.MatMulNBits op in a few key areas.

- The 'N' and 'K' attributes can be inferred from the input shapes of A and B inputs
- The 'bits' and 'block_size' attributes are optional with default values if not given
- The 'g_idx' group index input is removed. This only exists for specific hardware
- The int32 type is removed from the T2 tensor type. Unable to get a single use case that the T2 tensor type would be anything other than uint8.
- Fixes an error in the shape specified for the zero_points input when using uint8 datatype
- other small differences to the text to clarify how the N-Bit data is packed.

justinchuby commented:
Any consensus from the operators sig? Is this comment https://github.com/onnx/onnx/issues/6326#issuecomment-2318650903 relavent?

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,273 @@ ONNX_OPERATOR_SET_SCHEMA(
+
+    in which:
+      - `n_blocks_per_col` = `(K + block_size - 1) / block_size`
+      - `blob_size` = `CeilDiv(block_size * bits, bitsof(uint8_t)<8>)`
```
Can we replace `bitsof(uint8_t)<8>` by `8`? I am not sure I understand what is currently written.

georgen117 replied:
Yes, I can replace this with 8

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,273 @@ ONNX_OPERATOR_SET_SCHEMA(
+`[N * n_blocks_per_col]`
+
+Input `zero_points` is stored as `uint8_t` or the same type as `A`. It has the same packing method as input `B`.
+  - `[N * CeilDiv(n_blocks_per_col * bits, 8)]`
```
Can we replace the second part by `blob_size`, assuming it is the same?

georgen117 replied:
This is not the same as `blob_size` so cannot be replaced.

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,273 @@ ONNX_OPERATOR_SET_SCHEMA(
+            OpSchema::NonDifferentiable)
+        .TypeConstraint("T1", {"tensor(float)", "tensor(float16)"},
+                        "Constrain input and output type to float/half_float tensors.")
+        .TypeConstraint("T2", {"tensor(uint8)"}, "Constrain quantized weight types to uint8/int32.")
```
Is the `int32` a typo? Same question 2 lines below.

georgen117 replied:
Good catch. Yes, both are typos from an earlier draft of the code that was missed I will remove.

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,273 @@ ONNX_OPERATOR_SET_SCHEMA(
+3. Input `B`'s scale and zero point are specified by input scales and zero_points.
+
+    Input `B` is stored as uint8_t with shape: `[N][n_blocks_per_col][blob_size]` or
+    `[N][n_blocks_per_col * blob_size]`
```
Do we need to support both formats (2D and 3D)? Seems to me it would be simpler to standardize on one or the other. But if this is helpful for backward compatibility (I mean, for easier migration of existing models), I guess we could support both. It may still be useful to recommend one format or another (maybe we can eventually deprecate the other).

georgen117 replied:
The only model I have seen used the 3D format. The unit tests in ONNX Runtime for com.microsoft.MatMulNBits used the 2D format. I have no problem standardizing on the one format. The 3D format makes would make the most sense.

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,273 @@ ONNX_OPERATOR_SET_SCHEMA(
+              scales_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType) {
+            fail_type_inference("inputs are expected to have tensor type and output type should not be null.");
+          }
+
```
Suggest adding
```cpp
if (!hasNInputShapes(ctx, 3)) {
   return;
}
```
to protect against types of unknown shape/rank. A similar check (eg., `hasInputShape(ctx, 3)`) down below before accessing bias' shape.

georgen117 replied:
If I understand correctly you are suggesting that the code return early if we discover no optional inputs.

It is not a bad idea but I don't think that would be correct unless I move the code to propagate the ElemType and output shape. I will look into doing that. 

Am I properly understanding your suggestion?

gramalingam replied:
Also: adding unit-tests for the shape-inference would be helpful. You can do that [here](https://github.com/onnx/onnx/blob/main/onnx/test/shape_inference_test.py)

georgen117 replied:
I added the code as suggested as well calls to `hasinputshape(ctx, 3/4)` for the optional inputs `zero_points` and `bias`.

I also added `MatMulNBits` to the shape-inference unit test thanks for pointing that out to me.

shubhambhokare1 commented:
Will review this week!

georgen117 commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,272 @@ ONNX_OPERATOR_SET_SCHEMA(
+            "It needs to be a power of 2 and not smaller than 16",
+            AttributeProto::INT,
+            static_cast<int64_t>(128))
+        .Input(0, "A", "The input tensor, not quantized", "T1", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
```
@shubhambhokare1 Sorry I didn't respond to these 2 weeks ago when you left them. I missed the emails informing me of comments. I will try and address your comments this week.

georgen117 replied:
I am updating the text. Good catch.

georgen117 commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,272 @@ ONNX_OPERATOR_SET_SCHEMA(
+          if (a_shape.dim_size() != 2) {
+            fail_type_inference("Input A tensor of wrong rank.");
+          }
+          if (b_shape.dim_size() < 2 || b_shape.dim_size() > 3) {
```
Based on suggestion from @gramalingam I updated the text to just be the 3D shape.


I updated the text in the Operators.md and elsewhere but somehow missed commiting the change for defs.cc.  The "B" input is "A 3 dimensional input tensor of shape [N][n_blocks_per_col][blob_size], i.e. an N by 2 dimensional data blob containing the packed B bits".  The code is correctly checking for the 3D input shape but the text was still stating 2D or 3D. I will update the text for the B input.

georgen117 commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,272 @@ ONNX_OPERATOR_SET_SCHEMA(
+            static_cast<int64_t>(128))
+        .Input(0, "A", "The input tensor, not quantized", "T1", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
+        .Input(1, "B", "1 or 2 dimensional data blob", "T2", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
+        .Input(2, "scales", "quantization scales", "T1", OpSchema::Single, true, 1, OpSchema::NonDifferentiable)
```
Will do. Thanks for catching this.

georgen117 commented:
@gramalingam and @shubhambhokare1 thanks for your text/code review. I believe I have addressed all of your suggestions and comments if you feel I have not sufficiently addressed them please let me know.

liqunfu commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+            2,
+            "scales",
+            "Tensor of quantization scales. It should have shape [N * n_blocks_per_col]",
+            "T1",
```
I would not tie scales' type to the activation and output type (i.e. T1). To reduce model size, it is possible to use float16 for scales even when input/output type is float32. There is an example at https://github.com/ggerganov/llama.cpp/blob/f112d198cd9953b633ac662c2705d6e423438717/ggml/src/ggml-common.h#L213 where float16 is used and output is float32.

nzmora-nvidia commented:
I have misgivings about this PR that I've listed below, but my main point is that MatMulNBits is a tailormade operator that is not generic enough to serve the ONNX community at large.

1. Repetition: aside from n-bit support, this operator is a constrained composition of existing operators. I.e. it can be implemented using existing operators and assembled into a local function when desired. E.g. TensorRT implements weight-only-quantization using existing operators.

2. ONNX already has 4 matmul variants (matmul, matmulInteger, QLinearMatMul, GEMM) - do we need another one? Each operator added to ONNX has a cost incurred by many libraries for implementing, documenting, testing and maintaining.

3. MatMulNBits breaks the single-responsibility principle (SRP) and open-closed principle (OCP), both of which facilitate API simplicity and scalability. An operator either multiplies matrices or quantizes data - not both. Why? because the policy of each of these can change independently. For example: say that next year we MatMulNBits to support FP4 - we'll have to "open" MatMulNBits and change it (OCP).

4. Consistency: prior to the introduction of Q/DQ ONNX was a mess with specialized quantized operators. Then came Q/DQ to decouple the details of quantization from the details of the operation (SRP/OCP). MatMulNBits is inconsistent with the existing API and general direction of ONNX and will confuse some users.

5. MatMulNBits is very specialized and cannot serve many users. I've listed below some of the constraints this operator places on operands and behavior:
   - Supports n-bit integers but not n-bit floats
   - n-bit integers are not fully specified: signed? does it use 2's complement representation? What is the rounding policy?
   - Input B is a (i) quantized; (ii) 2D; (iii) constant
   - Input A is not quantized
   - How is an n-bit constant in ONNX without defining a type? What if B is transposed (often the case) - how can we represent a transpose on B which is an n-bit quantized constant?
   - blocking axis == 0, by definition
   - The shape of the scales depends on the types of the inputs (inconsistent with other operators).
   - "Input `scales` is stored in same type as original type of B(`float32`, `float16`) with shape like..." - How does the operator know the "original type of B" when we only see the quantized B?

georgen117 commented:
@nzmora-nvidia I wonder what your feelings are about the possibility of adding this as a function instead of an operator?

My original intent was to add this as a function but ran into two issues.
 1) No support for N-Bit data 
 2) The original com.microsoft.MatMulNBits specified the zero_points could be a floating type (`tensor(uint8), tensor(float16), tensor(float`) which is not supported by the current DequantizeLinear.

A new Operator could be added to supporting N-Bit data (for example UnpackNBits). Which is much more in line with single-responsibility principle (SRP) and open-closed principle (OCP). I don't really see a way to avoid adding  something for N-Bit data support to keep the N-Bits feature.

nzmora-nvidia commented:
Hi @georgen117 ,

Re 1: In the [opening comment](https://github.com/onnx/onnx/pull/6500#issue-2618862173) you named INT4 Phi3 as a motivator for the operator. Does INT4 suffice? If not, you might consider adding an n-bit integer to ONNX. I don't see how you can do without defining a new type because at minimum we would need a way to define n-bit (packed) constants.

Re 2: ONNX [defines ](https://onnx.ai/onnx/operators/onnx__DequantizeLinear.html) DQ as `real_value = scale * (quantized_value - zero_point) ` and under this definition `zero_point` must be quantized.

The equivalent definition: `real_value = scale * quantized_value - zero_point` - in which zero_point is a float - is considered biased  because it doesn't have an exact quantized representation of zero (`quantized_zero = round_and_clamp((zero_point / scale))`.
For details see [gemmlowp](https://github.com/google/gemmlowp/blob/master/doc/quantization.md

You could try a PR for DQ to support the bias definition as well.
Or maybe you can WAR with `DQ(x_q, scale=1) + zero_point)`

Regarding defining this as a function - why not as a local function? Does this operator appear in many models?

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+[MatMul](#matmul) with differences:
+
+1. Input `B` is a 2D constant Matrix. Its input feature count and output feature count are specified by attributes 'K'
+    and 'N'.
```
nit: this `N` is different from the `N` is first line (which is the number of bits in quantized type). May be change it to M? Also, change "input feature count" to "input feature count (number of rows)" and likewise for other dim.

georgen117 replied:
You are correct this N is the from and `MatMul((M,K) (K,N))` of the weight matrix. In the first line it is listed as N bits I could change the first line to N-Bits or NBits to help differentiate.  I think most people working with MatMul for 2D matrix multiplication are quite familiar with M, K, N matrix notation.

gramalingam replied:
That's fine, as long as the names are distinct, thanks.

gramalingam commented on onnx/defs/math/defs.cc:
````diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+        Byte 1: [ . . . .  C C C C] -> C[0:3]
+      ```
+
+    - for 3,5,6,7 bits, 32x3bit,32x5bit,16x6bit,32x7bit are stored in 12x`uint8_t`, 20x`uint8_t`, 12x`uint8_t`,
````
For clarification: so, this packing is different from what onnxruntime does? (May require repacking if we migrate models using the onnxruntime op to this op.)

georgen117 replied:
The 4-Bit integer packing is identical to the packing specified for int4 by ONNX specification and ONNX Runtime.

The packing done for all the other N-Bit types is specified this way to match the specification for int4. Although onnxruntime specifies the com.microsoft.MatMulNBits they currently only have an implementation for the 4-bit variant at this time. I have tested the packing as specified for 4-bit data in ONNX Runtime and get the correct results.

My first draft had the order different `[AAAABBBB]` for Byte1 and `[CCCC....]` for Byte2 found out though testing it was not packed as expected. That is when I found the [4-bit integer types](https://onnx.ai/onnx/technical/int4.html) link and matched the description to that.

gramalingam commented on onnx/defs/math/defs.cc:
````diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+      ```
+  The last `uint_8` byte may have some bits unused.
+
+Input `scales` is stored in same type as original type of B(`float32`, `float16`) with shape like:
````
With row-major-ordering (all entries for a given column appear consecutively), I assume?

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+        .Attr(
+            "accuracy_level",
+            "The minimum accuracy level of input A, can be: "
+            "0(unset), 1(float), 2(float16), 3(Bfloat16), 4(int8) (default unset). It is used to control how input A "
```
nit: I think it could help to use the same datatype encoding here as [TensorProto:DataType](https://github.com/onnx/onnx/blob/25a134a6e67885c859edf88d0f492430d2c4cd5a/onnx/onnx.proto#L504) ... eg., it could be directly used as the target type in a Cast op without converting.

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+        .Output(
+            0,
+            "Y",
+            "Matrix multiply results. The output tensor has the same rank as the input.",
```
Maybe say this is of shape [M][N] ? As I understand it, this supports only 2D inputs, without support for broadcast. So, no need for the second line about "same rank as the input".

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+            true,
+            1,
+            OpSchema::NonDifferentiable)
+        .TypeConstraint("T1", {"tensor(float)", "tensor(float16)"},
```
Is there any usage of bfloat16? General question, for other readers/reviewers as well ... sometimes we find requests to add support for bfloat16 to ops.

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+                                 "Shape B and N attribute don't match.");
+          }
+
+          int64_t n_blocks_per_col = (K + block_size - 1) / block_size;
```
In rare cases, K could be -1 here (unknown). We should skip corresponding checks involving n_blocks_per_col in that case.

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+                fail_shape_inference("Input uint8 zero_points dimensions is incompatible for MatMulNBits expected ",
+                                     "[N * CeilDiv(n_blocks_per_col * bits, 8)] which is [", bits_per_zero_points,
+                                     "] got [", zero_points_shape.dim(0).dim_value(), "].");
+              }
```
I think a check that `(zero_points_dtype == scales_dtype) || (zero_points_dtype == b_dtype)` could be added. I assume that is required? Eg., is it allowed for T1 to be float and T3 to be float16? 

Also: related to Liqun's point above: if we allow scale-type to be different, will need to change the conditions accordingly here.

gramalingam commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+
+          ONNX_NAMESPACE::TensorShapeProto resultShape;
+          *resultShape.add_dim() = a_shape.dim(0);  // M
+          resultShape.add_dim()->set_dim_value(N);  // N
```
Suggest replacing this by:
```cpp
auto* last_dim = resultShape.add_dim();
if (N != -1)
   last_dim->set_dim_value(N);
```

gramalingam replied:
ONNX does not use/recognize -1 as an unknown dim-value.

gramalingam commented:
> I wonder what your feelings are about the possibility of adding this as a function instead of an operator?

Wanted to bring this question up again. To make this work, can we restrict attention to 4-bit quantization, and define a function-op that takes the form "dequantize, matmul, quantize"?

I want to decouple the concerns expressed into two parts:
* Concerns specific to the details of the op (item 5 in @nzmora-nvidia 's message above).
* The other items in that message.

As a policy, ONNX supports the introduction of function-ops that serve to represent fusion-opportunities (which backends can easily ignore by expanding the function-op into more primitive ops).

I agree that community favors the QDQ pattern. However, there is value in defining function-ops that expand into the D-op-Q pattern. The point is that backends will not support all possible QDQ patterns, and it helps frontends to know which QDQ combinations are more likely to be supported.

saurabhtangri commented:
> > I wonder what your feelings are about the possibility of adding this as a function instead of an operator?
> 
> Wanted to bring this question up again. To make this work, can we restrict attention to 4-bit quantization, and define a function-op that takes the form "dequantize, matmul, quantize"?
> 
> I want to decouple the concerns expressed into two parts:
> 
> * Concerns specific to the details of the op (item 5 in @nzmora-nvidia 's message above).
> * The other items in that message.
> 
> As a policy, ONNX supports the introduction of function-ops that serve to represent fusion-opportunities (which backends can easily ignore by expanding the function-op into more primitive ops).
> 
> I agree that community favors the QDQ pattern. However, there is value in defining function-ops that expand into the D-op-Q pattern. The point is that backends will not support all possible QDQ patterns, and it helps frontends to know which QDQ combinations are more likely to be supported.

I think I can agree to above. The point of introducing decomposable ONNX functions was to ensure we don't inflate operator sets. If for ONNX functions we hold the same bar that was defined for ONNX operators then the core sentiment behind functions is diluted. We also need to make sure ONNX doesn't become stagnant and offers a fast/simpler path during model conversions.

saurabhtangri commented:
> > I wonder what your feelings are about the possibility of adding this as a function instead of an operator?
> 
> Wanted to bring this question up again. To make this work, can we restrict attention to 4-bit quantization, and define a function-op that takes the form "dequantize, matmul, quantize"?
> 
> I want to decouple the concerns expressed into two parts:
> 
> * Concerns specific to the details of the op (item 5 in @nzmora-nvidia 's message above).
> * The other items in that message.
> 
> As a policy, ONNX supports the introduction of function-ops that serve to represent fusion-opportunities (which backends can easily ignore by expanding the function-op into more primitive ops).
> 
> I agree that community favors the QDQ pattern. However, there is value in defining function-ops that expand into the D-op-Q pattern. The point is that backends will not support all possible QDQ patterns, and it helps frontends to know which QDQ combinations are more likely to be supported.

I can agree with the above The motivation behind ntroducing decomposable ONNX functions was to ensure we don't inflate ONNX operator sets. If for ONNX functions we hold the same bar that was defined for ONNX operators then the core sentiment behind functions is diluted. We also need to make sure ONNX doesn't become stagnant and continues to offer a fast/simple path during conversions.

saurabhtangri commented:
> Is there a clear document describing the concept and a guide on how to inline these ONNX functions from a backend perspective ?
> 
> I am not in favor of introducing this as an Operator as I agree with the above that the QDQ approach should be the preferred one. Have we explored a middle ground here ? Adding `MatMulNBits` as an ONNX function with a subgraph that "rounds" up to the next fitting integer dtype ? E.g. a `MatMulNBit` with 6 bits could be inlined to the pattern `DQ(int8) -> MatMul -> Q(int8)` as this should preserve accuracy. In addition this requires either being able to unpack the byte blob in the inliner or a `UnpackNBit` op. This should ideally be able to unpack using a sign, mantissa and exponent where integer unpacking sets the exponent bits to 0.
> 
> What is the plan to arrive at ONNX files with this operator ? Currently it is only produced by Intel Neural compressor, as ONNX Runtime GenAI already has the option to represent this as `QDQ+MatMul`. This arises the question for me if this requires implementation in ONNX spec or if this is an experimental op that will produce another ONNX data type like `int6` or `int2` if it becomes popular.

Yes we need to be resistive when adding more operators.

nzmora-nvidia commented:
> > Is there a clear document describing the concept and a guide on how to inline these ONNX functions from a backend perspective ?
> > I am not in favor of introducing this as an Operator as I agree with the above that the QDQ approach should be the preferred one. Have we explored a middle ground here ? Adding `MatMulNBits` as an ONNX function with a subgraph that "rounds" up to the next fitting integer dtype ? E.g. a `MatMulNBit` with 6 bits could be inlined to the pattern `DQ(int8) -> MatMul -> Q(int8)` as this should preserve accuracy. In addition this requires either being able to unpack the byte blob in the inliner or a `UnpackNBit` op. This should ideally be able to unpack using a sign, mantissa and exponent where integer unpacking sets the exponent bits to 0.
> > What is the plan to arrive at ONNX files with this operator ? Currently it is only produced by Intel Neural compressor, as ONNX Runtime GenAI already has the option to represent this as `QDQ+MatMul`. This arises the question for me if this requires implementation in ONNX spec or if this is an experimental op that will produce another ONNX data type like `int6` or `int2` if it becomes popular.
> 
> Yes we need to be resistive when adding more operators.

What are the differences in how ONNX ops and ONNX functions are treated at the spec/definition/formalism level?
1. An ONNX op is a formally-defined concept, with a definition, documentation, tests etc in the ONNX codebase.
2. An ONNX model-local function is self-defined and its scope is local to the model in which it was defined.
3. What's the deal with a (non-model-local) ONNX function? Does it need a formal and rigorous definition? Does it need a reference implementation? Does it Does it require tests? What makes it different from a regular Op (many ops already have a function definition) besides being optional?

liqunfu commented on onnx/defs/math/defs.cc:
```diff
@@ -2006,6 +2007,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+            AttributeProto::INT,
+            static_cast<int64_t>(0))
+        .Attr(
+            "bits",
```
the attribute "bits" alone is not enough to specify type of quantization. For example, with bitnet ternary quantization, it takes 2 bit only for 3 states of -1, 0 1. It may further compress the ternary weight to 3 weight values per 5 nits. I  suggest to add a optional attribute to allow indicating this types of cases. In terms of which string to use, we can discuss.
